### PR TITLE
Use emplace & emplace_back as more as possible

### PR DESCRIPTION
**Pull request by ckyang**
## Description

Use emplace() instead of push_back()/insert() can improve efficiency and reduce memory stack usage, because this API would not create a temporary object when copying object into container.

Following syntax needs to be refined:

unurdered_map::insert() -> unurdered_map::emplace()
vector::push_back() -> vector::emplace_back()
vector::insert() -> vector::emplace()
deque::push_back() -> deque::emplace_back()
deque::push_front() -> deque::emplace_front()
deque::insert() -> deque::emplace()

https://github.com/Zilliqa/Issues/issues/87

### Implementation

- [ ] implement ...
- [x] **ready for review**

### Integration Test (Core Team)

- [x] local machine test
- [x] small-scale cloud test

### CODE REVIEW

**Comment by bb111189** [comment]
Type casting can be remove here. 
https://github.com/Zilliqa/Zilliqa/blob/4510e5f69291b961f616320f24f1f7a274751642/src/libDirectoryService/FinalBlockPostProcessing.cpp#L340

**Comment by bb111189** [comment]
https://github.com/Zilliqa/Zilliqa/blob/4510e5f69291b961f616320f24f1f7a274751642/src/libNode/Node.cpp#L146

https://github.com/Zilliqa/Zilliqa/blob/4510e5f69291b961f616320f24f1f7a274751642/src/libLookup/Lookup.cpp#L1669

**Comment by ckyang** [comment]
This PR already merged into https://github.com/Zilliqa/Zilliqa/pull/378/, so I close this one.

**Comment by ansnunez** [comment]
Reopened branch and PR so we can review and commit this particular update separately from #378

File: src/libData/BlockData/Block/TxBlock.cpp [hunk 1]
```diff
@@ -116,7 +116,7 @@ int TxBlock::Deserialize(const vector<unsigned char>& src, unsigned int offset)
                 = GetNumber<uint32_t>(src, curOffset, sizeof(uint32_t));
             curOffset += sizeof(uint32_t);
 
-            m_shardIDs.emplace_back(shardID);
+            m_shardIDs.push_back(shardID);
```
Comment (ansnunez) [inline src/libData/BlockData/Block/TxBlock.cpp]: Any reason why this can't be emplace_back?

Comment (ckyang) [reply src/libData/BlockData/Block/TxBlock.cpp]: Please refer to this article: https://abseil.io/tips/112
There is always trade-off between performance & maintenance.
That is, for native data type, it doesn't matter to use push_back/emplace_back and we prefer to use push_back for good maintenance.
For user-defined data type, of course we prefer emplace_back for better performance.

**Pull request merged**